### PR TITLE
Made Nosync and ValidateModels modules thread-safe

### PR DESCRIPTION
**Pull request by kobsy**
### Summary
😱 

So this ended up being _slightly_ more involved than I'd anticipated. 😅 

There were a few issues that made this tricky. First was that some of the gems used for testing were _woefully_ out-of-date, to the point where it was locking our version of `activeresource` to 4.0, when it's running > 5.x in all of our Rails 5-based apps. 😬 I was able to sort that out, which allowed me to run the tests against updated Rails ... which revealed a _lot_ of failing tests.

I'm not sure if it's just the nature of updated minitest or what, but a number of the tests were unstable after updating -- even before I changed any of the thread-dangerous code! This was addressed by adding the `reset_nosync!` method in the second commit. That finally allowed the tests to be more deterministic, though they were still reliably failing on about 6 tests! 😓 

The ultimate culprit turned out to be a refactor in Rails 6 internals where part of the AR instantiation was extracted to a different method than the one we had hacked into. Hacking into the extracted method finally resolved the majority of the failing tests.

Then there was the actual matter of our `nosync` code. The trouble was that it was really being used in two ways: to `extend` the base AR class _and_ then also `include`d on the AR instances. 😱 My solution was to essentially copy the code into `InstanceMethods` and `ClassMethods` modules and include/extend appropriately. The instance methods use instance variables; the class methods use thread safe accessors.

It's kind of ugly, but it seems to work now, with passing tests and everything! 😅 

I'd welcome your feedback and thoughts. 🙇

### CODE REVIEW

File: remotable.gemspec [hunk 1]
```diff
@@ -15,11 +15,14 @@ Gem::Specification.new do |s|
   s.add_dependency "activerecord"
   s.add_dependency "activesupport"
 
+  s.add_development_dependency "appraisal"
```
Comment (lukebooth) [inline remotable.gemspec]: 👏 👏 👏 👏 
I've not heard of this until now. Very cool.

File: lib/remotable/nosync.rb [hunk 2]
```diff
@@ -1,37 +1,84 @@
...
+      def reset_nosync!
+        remove_instance_variable(:@nosync) if instance_variable_defined?(:@nosync)
+      end
+
+      def nosync(new_value=true)
+        old_value = _nosync
+        self.nosync = new_value
+        yield
+      ensure
+        @nosync = old_value
```
Comment (lukebooth) [inline lib/remotable/nosync.rb]: Could we make this assign happen via the method like the other lines do with `self.nosync=`? Or if there's a reason this `ensure` needs to directly assign to the instance variable, add a comment to say why? #forfutureus

File: lib/remotable/nosync.rb [hunk 3]
```diff
@@ -1,37 +1,84 @@
...
+      def nosync=(val)
+        @nosync = val
+      end
+
+      def nosync_value?
+        !_nosync.nil?
+      end
+
+      def nosync?
+        _nosync == true
```
Comment (lukebooth) [inline lib/remotable/nosync.rb]: ⬇️ The `== true` part is probably not necessary, but I'm also ok with only making changes necessary to the current code red issue 😅

File: remotable.gemspec (hunk 1 above)
Comment (kobsy) [reply remotable.gemspec]: I learned about it from the `pluck_map` gem. Seems pretty handy! 😁

File: lib/remotable/nosync.rb (hunk 2 above)
Comment (kobsy) [reply lib/remotable/nosync.rb]: I don't think there's necessarily a reason; I think this is what it _was_ doing before isolating it in the `InstanceMethods`, but I can see about just calling `self.nosync=`. 🤷‍♂

File: lib/remotable/nosync.rb (hunk 3 above)
Comment (kobsy) [reply lib/remotable/nosync.rb]: Yeah, I think we should be okay if we want to just refactor this to not use `== true`.

**Comment by kobsy** [comment]
I tried adding a commit that dedupes some of the nosync logic by implementing `ClassMethods` using `InstanceMethods`. Want to give it a look, @lukebooth?

**Comment by lukebooth** [comment]
Looks great! 🚀

**Pull request merged**